WordPress/gutenberg
# Components: refactor `Snackbar` to pass `exhaustive-deps`

## What?
Updates the `Snackbar` component to satisfy the `exhaustive-deps` eslint rule

## Why?
Part of the effort in https://github.com/WordPress/gutenberg/pull/41166 to apply `exhuastive-deps` to the Components package

## How?
Adds a single missing dependency (`explicitDismiss`) to the relevant dependency array.

## Testing Instructions
1. From your local Gutenberg directory, run `npx eslint --rule 'react-hooks/exhaustive-deps: warn' packages/components/src/snackbar`
2. Confirm that the linter returns no errors
3. Confirm unit tests still pass
4. Run Storybook locally, confirm the components stories and/or docs still work as expected, but there are some extra steps for Storybook testing on this component.

`Snackbar` doesn't have any real interaction in Storybook. The stories all lack an `onRemove` prop to handle the dismissal, presumably to keep the notices visible. They wouldn't be very good examples if they disappeared and left a blank screen 😉. We may want to consider including two matching (and clearly labeled) notifications in each story, keeping one visible (current behavior) and using the second to demonstrate what dismissing the notice does/looks like... but that would be for another PR.

Another relevant detail (also for another PR to address) is that the two "with `explicitDismiss`" stories are currently bugged in Storybook. This is because the stories aren't rendered by a `SnackbarList` component, so they don't have a `listRef` prop. The dismissal method `dismissMe` tries to use that prop to shift focus when the notice is dismissed, but currently the ref is `undefined` in Storybook, so we get a type error.

So... to actually test the interaction of the `explicitDismiss`, a few temporary changes are needed to make the actions of the `Snackbar` visible. Sorry, this is a bit of a pain.

**snackbar/index.js**
- line 20: shorten the `NOTICE_TIMEOUT` value to `1000`. Ten seconds is an _eternity_ to test with.
- comment out line 68. That ref doesn't exist in our stories, so it'll just cause an error.

**snackbar/stories.js**
- line 19: we need to simulate a dismissal, so declare a small logging function:
```
const onRemove = () => {
	console.log( 'Snackbar dismissed!' );
};
```
- pass the new `onRemove` function to the `onRemove` prop for each of the five stories in the file.

Now, to test the interactions:
- Starting with the default, click through each of the stories. Waiting at least one second each time.
- For the first three, confirm that our logging message appears in the console after the one second delay. This confirms that the function to dismiss the `Snackbar` automatically still works as expected.
- For the two `explicitDismiss` stories, confirm that the log does not automatically appear, confirming that these `Snackbar`s will persist until dismissed.
- Click the dismiss icon for each of the `explicitDismiss` stories, confirming our log message appears when clicked, which simulates a successful manual dismissal.
